gap/generic/kernel.gi

```diff
@@ -75,38 +75,46 @@ InstallGlobalFunction( RandomSubproduct, function(a)
 end );
 
 # Computes randomly (it might underestimate) the normal closure of <list>
-# under conjugation by the group generated by <grpgens>.
+# under conjugation by the group generated by <grpgens>. The integer
+# <n> controls how many new conjugates are computed.
+# The error probability can be determined by following Theorem 2.3.9 in
+# <Cite Key="Ser03"/>.  According to Lemma 2.3.8, if <grpgens> has 4 or more
+# entries, then as long as the result does not generate the full normal closure
+# of <list> under <grpgens>, the probability that a conjugate is not contained
+# in the group generated by the result is >= 1/4.
 InstallGlobalFunction( FastNormalClosure , function( grpgens, list, n )
-  local i,list2,randgens,randlist;
-  list2:=ShallowCopy(list);
-  if Length(grpgens) > 3 then
-    for i in [1..6*n] do
-      if Length(list2)=1 then
-        randlist:=list2[1];
-      else
-        randlist:=RandomSubproduct(list2);
-      fi;
-      if not IsOne(randlist) then
-        randgens:=RandomSubproduct(grpgens);
-        if not IsOne(randgens) then
-          Add(list2,randlist^randgens);
-        fi;
-      fi;
-    od;
-  else # for short generator lists, conjugate with all generators
-    for i in [1..3*n] do
-      if Length(list2)=1 then
-        randlist:=list2[1];
-      else
-        randlist:=RandomSubproduct(list2);
-      fi;
-      if not IsOne(randlist) then
-         for randgens in grpgens do
-             Add(list2, randlist^randgens);
-         od;
+  local list2, fewGenerators, repetitions, randlist, conjugators, i, c;
+  list2 := ShallowCopy(list);
+  if Length(grpgens) = 1 then
+    return list2;
```

Comment:
Note that if we instead of `gens` we passed a group object, we could test here for `HasIsAbelian(G) and IsAbelian(G)`. And in all places calling here that I found, there already was such a group object -- with the exception of a call in `IsThisSL2Natural`. But there in turn, all (?) callers to `IsThisSL2Natural` could easily be changed to pass on a group object.

One could even support both variants in this function, by turning `gens` into a local var, and calling the first argument `G`, and doing something like `if IsGroup(G) then gens := GeneratorsOfGroup(G); else gens:=G; fi;` or so

Reply:
I'll put this idea into its own issue.